src/app/tests/AppTestContext.h

```diff
@@ -36,12 +37,65 @@ class AppContext : public MessagingContext
     // Shutdown all layers, finalize operations
     CHIP_ERROR Shutdown();
 
+    /*
+     * For unit-tests that simulate end-to-end transmission and reception of messages in loopback mode,
+     * this mode better replicates a real-functioning stack that correctly handles the processing
+     * of a transmitted message as an asynchronous, bottom half handler dispatched after the current execution context has
+     completed.
+     * This is achieved using SystemLayer::ScheduleWork.
+
+     * This should be used on conjunction with the DrainAndServiceIO function below to correctly service and drain the event queue.
+     *
+     */
+    void EnableAsyncDispatch()
+    {
+        auto & impl = mTransportManager.GetTransport().GetImplAtIndex<0>();
+        impl.EnableAsyncDispatch(&mIOContext.GetSystemLayer());
+    }
+
+    /*
+     * This drives the servicing of events using the embedded IOContext while there are pending
+     * messages in the loopback transport's pending message queue. This should run to completion
+     * in well-behaved logic (i.e there isn't an indefinite ping-pong of messages transmitted back
+     * and forth).
+     *
+     * Consequently, this is guarded with a user-provided timeout to ensure we don't have unit-tests that stall
+     * in CI due to bugs in the code that is being tested.
+     *
+     * This DOES NOT ensure that all pending events are serviced to completion (i.e timers, any ScheduleWork calls).
+     *
+     */
+    void DrainAndServiceIO(System::Clock::Timeout maxWait = chip::System::Clock::Seconds16(5))
```

Comment:
Agree, we can have an interface like `DriveIOUntil(lambda)`.

Reply:
5 seconds for a unit test seems long to me - I would argue that our test should be expected to run quite fast. Waiting for CI error takes up to 1 hour or more.

Reply:
5s is actually quite a long time in terms of # of event loop iterations that can be run, since we're only blocking waiting on 0 pending messages outstanding (which in most cases, should happen within a handful of loop iterations, the above tests I profiled them with had loop counts of 4).

In the case of run-away conditions, I agree with @andy31415 that having it exit in 5s vs 1hr is a big win.